src/audio/pcm_converter/pcm_converter_generic.c

```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Tomasz Lauda <tomasz.lauda@linux.intel.com>
+
+/**
+ * \file audio/pcm_converter/pcm_converter_generic.c
+ * \brief PCM converter generic processing implementation
+ * \authors Tomasz Lauda <tomasz.lauda@linux.intel.com>
+ */
+
+#include <sof/audio/buffer.h>
+#include <sof/audio/pcm_converter.h>
+#include <sof/common.h>
+#include <ipc/stream.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static void pcm_copy_s16(struct comp_buffer *source, struct comp_buffer *sink,
+			 uint32_t samples)
+{
+	uint32_t buff_frag = 0;
+	int16_t *src;
+	int16_t *dst;
+	uint32_t i;
+
+	for (i = 0; i < samples; i++) {
+		src = buffer_read_frag_s16(source, buff_frag);
+		dst = buffer_write_frag_s16(sink, buff_frag);
+		*dst = *src;
```

Comment:
Have you considered memcpy() based function instead? There's example in src_copy_s16()?

Reply:
@singalsu What about "non-uniform wrapping", as I would call? For example we want to copy 20 samples. After 10 samples the source buffer wraps, after another 5 the destination buffer wraps. I think this code, even if it may be less performant (I'm not very convinced it is, for the record) is simpler to understand.

Reply:
1:1 copy functions could be made available to library to avoid duplication. SRC uses it for same rate source & sink, and EQs use such for bypass mode.